lighthouse-core/audits/treemap-data.js

```diff
@@ -0,0 +1,315 @@
+    // of the source path, creating nodes as necessary.
+    for (const [source, data] of Object.entries(sourcesData)) {
+      addNode(source || `<unmapped>`, data, rootNode);
+
+      // Apply the data to the rootNode.
+      rootNode.resourceBytes += data.resourceBytes;
+      if (data.unusedBytes) rootNode.unusedBytes = (rootNode.unusedBytes || 0) + data.unusedBytes;
+    }
+
+    /**
+     * Collapse nodes that have only one child.
+     * @param {Node} node
+     */
+    function collapse(node) {
+      while (node.children && node.children.length === 1) {
+        node.name += '/' + node.children[0].name;
+        node.children = node.children[0].children;
+      }
+
+      if (node.children) {
+        for (const child of node.children) {
+          collapse(child);
+        }
+      }
+    }
+    collapse(rootNode);
+
+    // TODO(cjamcl): Should this structure be flattened for space savings?
+    // Like DOM Snapshot.
+    // Less JSON (no super nested children, and no repeated property names).
+
+    return rootNode;
+  }
+
+  /**
+   * Returns a root node where the first level of nodes are script URLs.
+   * If a script has a source map, that node will be set by prepareTreemapNodes.
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<RootNode[]>}
+   */
+  static async makeJavaScriptRootNodes(artifacts, context) {
+    /** @type {RootNode[]} */
+    const rootNodes = [];
+
+    const bundles = await JsBundles.request(artifacts, context);
+    const duplication = await ModuleDuplication.request(artifacts, context);
+
+    let inlineScriptLength = 0;
+    for (const scriptElement of artifacts.ScriptElements) {
+      // No src means script is inline.
+      // Combine these ScriptElements so that inline scripts show up as a single root node.
+      if (!scriptElement.src) {
+        inlineScriptLength += (scriptElement.content || '').length;
+      }
+    }
+    if (inlineScriptLength) {
+      const name = artifacts.URL.finalUrl;
+      rootNodes.push({
+        name,
+        node: {
+          name,
+          resourceBytes: inlineScriptLength,
+          unusedBytes: 0,
+        },
+      });
+    }
+
+    for (const scriptElement of artifacts.ScriptElements) {
+      if (!scriptElement.src) continue;
+
+      const name = scriptElement.src;
+      const bundle = bundles.find(bundle => scriptElement.src === bundle.script.src);
+      const scriptCoverages = artifacts.JsUsage[scriptElement.src];
+      if (!bundle || !scriptCoverages) {
+        rootNodes.push({
+          name,
+          node: {
+            name,
+            resourceBytes: scriptElement.src.length,
+          },
+        });
+        continue;
+      }
+
+      const unusedJavascriptSummary = await UnusedJavaScriptSummary.request(
+        {url: scriptElement.src, scriptCoverages, bundle}, context);
+
+      let node;
+      if (!unusedJavascriptSummary.sourcesWastedBytes) {
```

Comment:
can we get a refresher comment for these ifs on what cases they represent (and flip the negated one to the else?)